File: 2020/2.kt
// Advent of code 2020 day 2
import java.io.File

fun main() {
  val inputfile = "input2.txt"
  val lines: List<String> = File(inputfile).readLines()

  var valid1 = 0
  var valid2 = 0

  for (line: String in lines) {
    val parts: List<String> = line.split(" ") //.toList()
    val rule = parts[0]
    val letter: Char = parts[1].replace(":", "").single()
    val password = parts[2]

    val constraints: List<String> = rule.split("-")
    val lmin: Int = constraints[0].toInt()
    val lmax: Int = constraints[1].toInt()
    val count: Int = password.filter{ it == letter }.count()
    // println("$lmin, $lmax, $count, $letter, $password")

    // Part 1
    if (count >= lmin && count <= lmax) {
      valid1 += 1
    }

    // Part 2
    if ((password[lmin - 1] == letter) xor (password[lmax - 1] == letter)) {
      valid2 += 1
    }
  }

  println("$valid1, $valid2")
}

File: 2020/2.py
#!/usr/bin/env python

inputfile = "input2.txt"
with open(inputfile, "r") as f:
  lines = [x.strip() for x in f.readlines()]

"""
lines = [
  "1-3 a: abcde",
  "1-3 b: cdefg",
  "2-9 c: ccccccccc"
]
"""

valid1 = 0
valid2 = 0
for line in lines:
  rule, letter, password = line.split()
  lmin, lmax = [int(x) for x in rule.split("-")]
  letter = letter.strip(":")
  count = password.count(letter)
  print lmin, lmax, count, letter, password

  # Part 1
  if count >= lmin and count <= lmax:
    valid1 += 1

  # Part 2
  if (password[lmin - 1] == letter) ^ (password[lmax - 1] == letter):
    valid2 += 1

print valid1, valid2
